src/components/actions/processTrigger.ts

```diff
@@ -0,0 +1,101 @@
+import { ActionInTemplate, TriggerPayload, ActionSequential } from '../../types'
+import DBConnect from '../databaseConnect'
+import { actionLibrary } from '../pluginsConnect'
+import { EvaluatorNode } from '@openmsupply/expression-evaluator/lib/types'
+import { executeAction } from './executeAction'
+import { ActionQueueStatus, TriggerQueueStatus } from '../../generated/graphql'
+import { swapOutAliasedActions } from './helpers'
+
+// Dev config
+const showActionOutcomeLog = false
+
+export async function processTrigger(payload: TriggerPayload) {
+  const { trigger_id, trigger, table, record_id, data, event_code } = payload
+
+  const templateId = await DBConnect.getTemplateIdFromTrigger(payload.table, payload.record_id)
+
+  // Get Actions from matching Template (and match templateActionCode if applicable)
+  const actions = (await DBConnect.getActionsByTemplateId(templateId, trigger))
+    .filter((action) => {
+      if (!event_code) return true
+      else return action.event_code === event_code
+    })
+    .map((action) => (action.code !== 'alias' ? action : swapOutAliasedActions(templateId, action)))
+
+  const resolvedActions = await Promise.all(actions)
+
+  // Separate into Sequential and Async actions
+  const actionsSequential: ActionSequential[] = []
+  const actionsAsync: ActionInTemplate[] = []
+  for (const action of resolvedActions) {
+    if (action.sequence) actionsSequential.push(action as ActionSequential)
+    else actionsAsync.push(action)
+  }
+
+  for (const action of [...actionsAsync, ...actionsSequential]) {
+    // Add all actions to Action Queue
+    await DBConnect.addActionQueue({
+      trigger_event: trigger_id,
+      trigger_payload: payload,
+      template_id: templateId,
+      sequence: action.sequence,
+      action_code: action.code,
+      parameter_queries: action.parameter_queries,
+      parameters_evaluated: {},
+      condition_expression: action.condition,
+      status:
+        typeof action.sequence === 'number'
+          ? ActionQueueStatus.Processing
+          : ActionQueueStatus.Queued,
+    })
+  }
+  await DBConnect.updateTriggerQueueStatus({
+    status: TriggerQueueStatus.ActionsDispatched,
+    id: trigger_id,
+  })
+
+  // Get sequential Actions from database
+  const actionsToExecute = await DBConnect.getActionsProcessing(templateId)
```

Comment:
I've just been looking at this with fresh mind, and would be great to have a little comment here that asynchronous (queued) actions are initiated via pg notify

Reply:
Yeah good point, I've found myself wondering where the heck they get dealt with when looking through this before lol 850ae0fa